### PR TITLE
Initial implementation and tests

### PR DESCRIPTION
**Pull request by RangerMauve**
Fixes # .

Changes proposed in this pull request:
-
-
-

@consento-org/maintainer

### CODE REVIEW

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: `hasResponded` is set but not used.

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: Shouldn't this be always `n-1`?

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: This should probably be `state === "ready"` ?

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: When I first read `isWho` I was pretty confused, same for `WhoRequest` Maybe rename?

File: test/index.ts
Comment (martinheidegger) [inline test/index.ts:128]: There are a few essential tests missing here:

1. one member does an operation while another removes that member: who wins?
2. Multiple requests processed at once where the first adds a member (preventing the second to go through because more request are needed)
3. two members start an add operation in parallel, what happens?

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: Ugh, endless recursion possible, right? 😅

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: There should be a validate template here to validate if the feed data that is added keeps being added. I think this is important to figure out even in an early stage because it allows us to figure out if we **can** identify a valid from a corrupt feed.

File: test/index.ts
Comment (martinheidegger) [inline test/index.ts]: This adds the requirement that the other member knows the initiator at start. QR code for `member` has to be given to `other` and for `requestAdd` the `other`'s QR-code has to be given to `member`. It should be possible to reduce this so that only the `requestAdd(other.id)` is necessary ?!

File: test/index.ts
Comment (martinheidegger) [inline test/index.ts]: Shouldn't processFeeds() be part of the syncing process? What advantage do we get of separating those?

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: Memo: This needs to be established in the final solution using a log statement.

File: src/member.ts
Comment (martinheidegger) [inline src/member.ts]: This should work without the casting.

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Is there a way to get typescript to detect stuff like that? I feel like I might have to add eslint into the mix.

Having three tools to do linting feels a bit daunting. 😂 Parsing all the code at least 3 times per lint.

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: If we do `n-1` can we avoid ambiguity? Like, if we have A,B,C,D,E, if A,B,C have signed, and D/E aren't connected to each other / accept, they'll both think they're the final needed signature and cause a fork.

File: test/index.ts
Comment (RangerMauve) [reply test/index.ts:128]: Adding TODOs to the README to keep track of these, will work on them this week

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Err, what's the condition that causes it?

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Would this require the signvector stuff?

File: test/index.ts
Comment (RangerMauve) [reply test/index.ts]: My assumption was that only known members were allowed to request an add, therefore if a new peer was to be added, they'd need to know the ID of the first member in order to traverse the history and get to the latest state.

Ties into what we talked about regarding using just member feeds for representing history and traversing them instead of a group feed.

File: test/index.ts
Comment (RangerMauve) [reply test/index.ts]: Main thing here was the auto-accept function meant that processing feeds would trigger member adds, and I was hoping to have more control of when that happened. I can refactor to work around that so that control of accepting could be outside of syncing feeds.

Other consideration was that I wanted a place for members to process requests and responses that they generated themselves 1. for the initial member case and 2. to remove the need for extra logic for that bit. Down to add proccesFeeds at the end of sync, though.

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Could you elaborate on this point?

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Thank you. Still getting used to what is and isn't limited by TS. 😅

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: The variable was supposed to be used later. Good catch!

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Renamed

**Comment by RangerMauve** [comment]
Working on refactoring to change how processFeeds works, and then I'll add more tests

File: src/member.ts
Comment (dkastl) [reply src/member.ts]: What do you need Typescript for to detect this if you have Martin? ;-)

Comment (martinheidegger) [reply src/member.ts]: Note: my IDE uses code colors to highlight unused variables.

File: src/member.ts
Comment (martinheidegger) [reply src/member.ts]: If `n-1` doesn't work in that case it doesn't work in the other case, as two devices can mutually remove each other, it would also create that ambiguity.

File: src/member.ts
Comment (martinheidegger) [reply src/member.ts]: Nothing at the moment, but my eyes look with stress 💦 on the code and I would be constantly worried if it introduces endless recursions.

File: src/member.ts
Comment (martinheidegger) [reply src/member.ts]: The sign-vectors are part of this too. Yes, but there is a much simpler case: if there is a feed of a member that has been removed or was never added. Also, there is a thin possibility that the sign-vectors fail (and a fork is created). i.e. malfunctioning of a client: Fork detection should also be part of it.

Note: The central log may be a tool to make it easier/possible to detect member-feed forks.

File: src/member.ts
Comment (martinheidegger) [reply src/member.ts]: Setting a `finished = true` statement works for this simulation but in an actual case we should probably have a log statement that looks like `{ request: 'abcd', finished: true }` in order to recover the state from the log itself.

File: test/index.ts
Comment (martinheidegger) [reply test/index.ts]: > need to know the ID of the first member

The ID of the first member should be in the first _(and only?)_ block of the group feed, no?

But beside that: The workflow needs indeed to be `group-id` → new member → `new-member-id` → existing member.

**Comment by martinheidegger** [comment]
Without having a central feed its quite challenging for me to figure out _if_ a fork happened. Is there a good way to verify that without a central log?

Another thing: When adding feeds, the information should not just contain their ID's but also the verification vector for the first block

File: src/member.ts
Comment (martinheidegger) [reply src/member.ts]: We had a conversation about allowing forks but maybe resolving equal forks?! In example: D & E approve R1 creating Forks R1D and R1E and if they sync and notice equality they decide for the alphabetically lower one (R1D).

File: test/index.ts
Comment (martinheidegger) [reply test/index.ts]: > I can refactor to work around that so that control of accepting could be outside of syncing feeds.

Structurally it is probably for the best, though I don't think we need to use manual confirmation in our first real scenario. The reason why I brought this up is because in the sync process it may be possible to avoid some chance of forks by having one member sync first and process the entries before syncing everything - including the need entries - with the other peer.

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Is there a different approach I should take here? The current structure recurses until all the feeds have been processed (more specifically, until no items are processable).

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Hmm, mind pseudocodeing how the verification would look and what the test for forking would look like?

File: test/index.ts
Comment (RangerMauve) [reply test/index.ts]: Hmm, yeah, my current code doesn't account for the group feed or how the initial member is found. Was hoping to punt that for later and focus on the request flow.

File: test/index.ts
Comment (RangerMauve) [reply test/index.ts]: Yeah, I'm thinking that I'll have a `getPendingRequests()` method and a `acceptPendingRequests()`

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Hmm, not 100% sure what you mean regarding recovering state from the log. If we don't persist request state anywhere we'd need to traverse the entire log again and that means that `finished` would be derived from there again.

**Comment by RangerMauve** [comment]
Without a central feed there _is_ no forking.

Since everyone needs to agree on a history and you derive members by traversing each others feeds and verifying that signatures happened. If someone tires to fork their own member feed (assuming they bypass the signvector layer of security) then at most they can prevent new peers from getting a valid history (until the fork is detected and the invalid feed is purged).

Agreed on the verification vector stuff. I'll look at incorporating sign vectors after this initial refactor.

File: src/member.ts
Comment (RangerMauve) [reply src/member.ts]: Yeah, I usually have eslint cover unused variables and set up my tests to make sure. 😁

I guess eslint will be necessary after all. In which case I think I'd rather remove prettier from the workflow.

Comment (RangerMauve) [reply src/member.ts]: It'd be nice if the typescript type checking was inside eslint, too. I'll see if that's an option. 😁

File: test/index.ts
Comment (martinheidegger) [reply test/index.ts]: It is okay to make this a memo.

**Comment by RangerMauve** [comment]
TODO:

- Add sign vectors
- Verify sign vectors in sync
- When removing specify index to remove at so that they won't be processed after
- Add tests for forks
- Add tests fo rconflicting requests
- Add tests so that external peers always process feeds in the same way

File: src/member.ts
Comment (martinheidegger) [reply src/member.ts]:
```javascript
function isIncomingFeedForked (localFeed, feed) {
  // Look at the merkle tree and see if the feed was forked at a previous step
}

function isIncomingFeedSignVectorCorrect (feed, startIndex) {
  let vector = createSignVectorFor(feed.get(startIndex))
  for (let i = startIndex+1; i<feed.length; i++) {
    if (!signVector.validate(feed.get(i))) return false
  }
  return true
}

function isIncomingFeedCorrect (feed) {
  // testing "correct-ness" of requests, i.e. proper msgpack/protobuf object; size limit not exceeded; removes feed that actually existed. Does the signature exist for the request? etc.
}

function verifyFeed (localFeed, feed, startIndex, peer) {
  if (isIncomingFeedForked(localFeed, feed)) {
    // Red alert
    requestRemoveMember({ cause: 'fork-detected' })
    return false
  }
  if (!isIncomingFeedSignVectorCorrect(feed, startIndex)) {
    addToBlackList(peer)
    return false
  }
  if(!isIncomingFeedCorrect(feed)) {
    // orange alert
    requestRemoveMember({ cause: 'incorrect-client' })
    return false
  }
  return true
}
```

Comment (RangerMauve) [reply src/member.ts]: Sweet, thank you!

**Comment by RangerMauve** [comment]
Added some sims, so far so good.

Talking with Martin last week, we determined we could be running into issues if we allow members to have multiple in-flight requests.

Gonna write up some thoughts on how that works and make some diagrams of what it would look like.

**Comment by martinheidegger** [comment]
When adding https://github.com/consento-org/permissions-system-sim/pull/1/commits/4651ac0925e24363bee028a12b5dc6a001b558e7 it came to me that the one and two members situation not only is special, but also that the we have at least three options on how to deal with these.

1. Don't allow two members: The first member **has** to add two members at once. In practice this would mean having a virtual member that is basically a QR code.
2. Treat one member (either first or second) with higher power, making it possible for the stronger to remove the weaker but not the other way around.
3. (current) Require both members to sign the removal of one. Meaning that you can't "loose" either of the initial two members.

**Comment by martinheidegger** [comment]
To figure out the complexity of the system I would like to propose a "possible action" api:

```javascript
const a = new Member()
const b = new Member(a.groupId)
a.possibleActions // [sealGroup, addMember]
b.possibleActions // [] - not a member of the group
a.possibleActions.addMember(b.id)

a.possibleActions // [syncB, requestAdd, revertAdd - as its not synced with B yet]
a.possibleActions.syncB()

a.possibleActions // [removeB, addMember, syncB]
// etc.
```

By having a list of "possible actions" we can create an operational matrix. Create branch test after each operation where we see if all possible options on each member are exhausted by the tests.

**Comment by RangerMauve** [comment]
> When adding [4651ac0](https://github.com/consento-org/permissions-system-sim/commit/4651ac0925e24363bee028a12b5dc6a001b558e7) it came to me that the one and two members situation not only is special, but also that the we have at least three options on how to deal with these.
> 
> 1. Don't allow two members: The first member **has** to add two members at once. In practice this would mean having a virtual member that is basically a QR code.
> 2. Treat one member (either first or second) with higher power, making it possible for the stronger to remove the weaker but not the other way around.
> 3. (current) Require both members to sign the removal of one. Meaning that you can't "loose" either of the initial two members.

Is the current really requiring both members to sign? I thought that removes were `n - 1`.

Personally I don't think adding two members at once is going to work well for UX.

I really like option 2 and think that "order that they were added in" would be a good way to rank power. I think that rather than stronger or weaker, it'd be like "who's been here longer". Though I can totally see the case where the older member will allow themselves to be removed through inaction. 🤷 But I guess that's an interesting case for forked feeds.

On another note,, I'm not sure I understand the possibleActions yet. What sorts of values can they have and when will those values change? The thing throwing me off is `syncB` Does that mean "sync request B"?

**Comment by RangerMauve** [comment]
BTW: Flow charts: https://hackmd.io/xiiC_CmxS1aV_aHO5qR4pw?both

Side note, in the comments on the flowcharts @martinheidegger  mentioned "How does Bob know Alice".

I was imagining that when the group is being created, the initial member would somehow get encoded into that groups identifier. That way you'd be able to start traversing the history of the initial member and find all the subsequent ones through adds.

Would that make sense? What if the group ID was the initial member ID even?

Side note, what should we do when we detect forks when there's only one peer online that can vouch that there's a different history available from what the forker provided? I think this is mostly a vulnerability with the initial member. 🤔

**Comment by martinheidegger** [comment]
> Is the current really requiring both members to sign? I thought that removes were n - 1.

I added this exception in https://github.com/consento-org/permissions-system-sim/pull/1/commits/4651ac0925e24363bee028a12b5dc6a001b558e7. If there are only two members: both are required. 

> Personally I don't think adding two members at once is going to work well for UX.

Adding two members would look like: you try to add a second member and at the same time you need to create a "virtual member" which is an additional password/secret key/qr-code that needs to be virtually restored in order to remove one of the two members.

> On another note,, I'm not sure I understand the possibleActions yet. What sorts of values can they have and when will those values change? The thing throwing me off is syncB Does that mean "sync request B"?

With `possibleActions` we could track if our tests actually tested all possible combinations. Action `A` causes `x` possible Actions per `y` other members and in the end we can track if we tested out all paths in the test somehow.

They could also be part of a "verification" tool: we only accept requests and signatures if they are actually "possible".

**Comment by RangerMauve** [comment]
Who would then control the virtual member? How would that be stored in any way that's different from either one member having priority on removes or needing both members to consent to a removal?

Is possibleActions kinda like the state checking code that we have now but a bit more declarative?

My gut there feeling is that defining it through actions will quickly grow the number of things that need to be checked vs something like `b.isMember` and `b.seesRequest('b')`. I don't feel super strongly about it though. 😁

**Comment by martinheidegger** [comment]
Expanding on my point:

The test framework could do autobranching:

```javascript
const executedOps = new Set()

function testPossibility (members, maxMembers, operation) {
  members = cloneMembers(members) // Cloned state
  validResult(operation(members, maxMembers)) // lets not exceed a max complexity here
  validateState(members) // Are the members (still) in a valid state, no forks?
}

function testPossibilities(members, maxMembers) {
  const possibleOps = combinePossibilities(members)
  for (const op of possibleOps) {
     if (executedOps.has(op.id)) {
       continue
     }
     executedOps.add(op.id)
     testPossibility(members, maxMembers, op)
  }
}

testPossibilities([new Member()], 7)
```

**Comment by RangerMauve** [comment]
Ah interesting. So the possibilities would try to generate every possible case and detect conflicting states or locks?

**Comment by martinheidegger** [comment]
> Who would then control the virtual member? How would that be stored in any way that's different from either one member having priority on removes or needing both members to consent to a removal?

The UX can still look similar as in the case 2.) the difference would just be that it gives the control and responsibility to the implementer. The implementer could make this in a way that this virtual member (which could be created in any case!) is simply stored at the same place as the first member or the second. Making the 1st member or the 2nd member stronger, depending on where that information is stored.

**Comment by martinheidegger** [comment]
> Ah interesting. So the possibilities would try to generate every possible case and detect conflicting states or locks?

Yes.

**Comment by RangerMauve** [comment]
@martinheidegger Regarding the virtual member, is this so that we can allow flexability to leave it to the app to choose which member has precedence?

Could the same thing be done with a flag in the add request for the second member? A member is a lot of state to keep track of and I feel like it'd make it a lot harder to implement in the client side (does a virtual member have a new replication swarm and ID?)

Also forcing there to be two members at once seems to add a bit of complexity. Are they a new type of request that has a `who` array instead of a `who` field? Is it two requests? How would that play with the one request per peer stuff.

What I'm trying to get at is that if we really need that flexibility there might be simpler ways to get it. 😅

**Comment by RangerMauve** [comment]
Regarding `possibleActions` I think that's a good idea, though it might be good to have some hand coded cases too so that we can get a better grasp on the details of what's happening. Testing all possible outcomes feels like a stress testing thing to have once the code is more stable rather than early experimental parts which can change drastically.

**Comment by martinheidegger** [comment]
> Could the same thing be done with a flag in the add request for the second member? A member is a lot of state to keep track of and I feel like it'd make it a lot harder to implement in the client side (does a virtual member have a new replication swarm and ID?)

Could be done with a flag, the point about the flag is that we need to add a <if> statement while the the other approach would push that responsibility to the user. _Note: the virtual member behaves and acts just like a regular member in our system._

> Also forcing there to be two members at once seems to add a bit of complexity. 

The other option would be that the creation of a group already requires 3 members to start off and can never have less than three. (First block contains 3 members instead of 2). Removing one member of a 3 member group could be equal to "close group".

**Comment by RangerMauve** [comment]
I'm still not sure what the benefit of pushing complexity onto users is here. 😅 It feels like it'd lead to the same dynamics as having two users even if there's an extra virtual user.

Needing a minimum of 3 members also feels like a rough spot for UX since I could imagine wanting to be able to use the system on a single device / add my other devices as I set them up. Or are we saying consento shouldn't be used there and should only be used for cases where you already have a group?

**Comment by martinheidegger** [comment]
Yeah, i cant really defend that option anymore.

**Comment by RangerMauve** [comment]
Issue I'm trying to think through right now is members resolving to different orders for knownMembers due to concurrent requests, and how that might be different from what external members would see. 😵

This'd be a problem if we use "who was first" for precedence in removes.

**Comment by RangerMauve** [comment]
Here's a diagram describing the problem. https://hackmd.io/xiiC_CmxS1aV_aHO5qR4pw?both#Concurrent-Requests---Problem

Ideally it'd be good to find a way to get members that resolve to something other than the "canonical" ordering to recover. Maybe they need to re-calculate after seeing a new accepted request?

**Comment by RangerMauve** [comment]
Just added a test that demonstrates the issue:

A is waiting for E to sign Request D
C is waiting for D to sign Request E
E and D think everything is fine and converge to the same set where they're both visible.

Is this where we should star thinking about request merges?

**Comment by martinheidegger** [comment]
One way how I could see this being resolved: Order members by ID in case of ambiguity (It is possible to identify the ambiguity)

**Comment by RangerMauve** [comment]
HMMMM. I think I'm going to sleep on it for now.

Some thoughts:

- Detecting the ambiguity we detect concurrent requests would be good. (test shows they all see 2 pending requests at some point)
- Would it make sense to try to enforce the order that E and D come to?
- When making a request, could we use a vector clock and the set of members that need to sign it to ensure stuff like this?

**Comment by martinheidegger** [comment]
> I think I'm going to sleep on it for now.

Sleep well!

> Would it make sense to try to enforce the order that E and D come to?

Well: resolving the ambiguity is in a sense "enforcement", isn't it?!

> When making a request, could we use a vector clock and the set of members that need to sign it to ensure stuff like this?

More information will make it easier for us to sort (i.e. we could use the accept order of all members) and it will clear up many ambiguities, but I am not sure if we can remove all of them.

**Comment by RangerMauve** [comment]
- Integrate Martin's HLC code
- Add timestamps to each Request and Response
- Sync timestamps with peers?
- When seeing a request, check what the members are at that time and add to the request time

**Comment by RangerMauve** [comment]
Added the clock, and did some refactoring.

Here's the main behavior that's changed:

- Requests and Responses are tagged with HLC clocks
- Responses get added to a list when they've been `finished`
- `knownMembers` is dynamically calculated from the finished responses (sorted by the time they were finished using HLC clocks)
- Requests are tagged with the knownMembers at the time they were seen (will change)
- If there are two concurrent requests, they will both be satisfied when the when the knownMembers at the time sign off.

Some more stuff to do:

- Validate there aren't any backwards time jumps in people's feeds
- Instead of passing knownMembers at the time, pass a getter so it can be generated as new requests get synced.

File: src/Permissions.ts
Comment (martinheidegger) [inline src/Permissions.ts]: We could use a Timestamp.compare util here.

File: src/Permissions.ts
Comment (martinheidegger) [inline src/Permissions.ts]: This maybe should be precomputed because this seems like a common operation?

File: src/Permissions.ts
Comment (martinheidegger) [inline src/Permissions.ts]: I read this as "When the request was opened" not as the "time of the open request".

File: src/Permissions.ts
Comment (martinheidegger) [inline src/Permissions.ts]: Why is this not sorted by time?

File: src/Permissions.ts
Comment (martinheidegger) [inline src/Permissions.ts]: TODO is done ... can be removed.

File: src/Permissions.ts
Comment (martinheidegger) [inline src/Permissions.ts]: It seems like this logic doesn't consider a removed member that was added previously?

File: src/Sync.ts
Comment (martinheidegger) [inline src/Sync.ts:48]: Use error codes.

File: src/VersionedStates.ts
Comment (RangerMauve) [inline src/VersionedStates.ts:88]: Why does this need to be placed into an array instead of using `yield`?

File: src/Permissions.ts
Comment (RangerMauve) [reply src/Permissions.ts]: What advantage does that have over using `time1.compare(time2)`?

File: src/Permissions.ts
Comment (RangerMauve) [reply src/Permissions.ts]: FN got replaced with versionedstate stuff

File: src/Permissions.ts
Comment (RangerMauve) [reply src/Permissions.ts]: Resolved by using versionedsate

File: src/Permissions.ts
Comment (RangerMauve) [reply src/Permissions.ts]: Renamed, how does `latestRequestTimestamp` sound?

File: src/Permissions.ts
Comment (RangerMauve) [reply src/Permissions.ts]: Now gets pre-computed

File: src/Permissions.ts
Comment (RangerMauve) [reply src/Permissions.ts]: Either way it's removed now. :P

File: src/VersionedStates.ts
Comment (martinheidegger) [reply src/VersionedStates.ts:88]: Sorting.

File: src/Permissions.ts
Comment (martinheidegger) [reply src/Permissions.ts]: Oh, right: it doesn't have any advantage: My bad. sorry for confusion

File: src/MemberList.ts
Comment (martinheidegger) [inline src/MemberList.ts]: As it is now: there is no precalculation happening. You should be able to just use byState and it will always be correct (trade: code brevity for performance)

File: src/VersionedStates.ts
Comment (RangerMauve) [reply src/VersionedStates.ts:88]: Ah, is that so we can iterate from oldest to newest? `iterateWithTime` makes sense as a name.

File: src/Sync.ts
Comment (RangerMauve) [reply src/Sync.ts:48]: Can we merge this in before doing the error codes refactor?

**Review by martinheidegger — APPROVED** [review]
We can merge it for now. more progress would be good to be made in issues.

**Pull request merged**
